src/Microsoft.Identity.Web.TokenAcquisition/TokenAcquisition.cs

```diff
@@ -585,10 +608,75 @@ private bool IsInvalidClientCertificateOrSignedAssertionError(MsalServiceExcepti
                     _applicationsByAuthorityClientId.TryAdd(GetApplicationKey(mergedOptions), application);
                 }
             }
+            return application;
+        }
 
+        /// <summary>
+        /// Gets a cached ManagedIdentityApplication object or builds a new one if not found.
+        /// </summary>
+        /// <param name="mergedOptions"></param>
+        /// <param name="managedIdentityOptions"></param>
+        /// <returns></returns>
+        internal async Task<IManagedIdentityApplication> GetOrBuildManagedIdentityApplication(
+            MergedOptions mergedOptions, ManagedIdentityOptions managedIdentityOptions)
+        {
+            IManagedIdentityApplication? application;
+            string key = GetCacheKeyForManagedId(managedIdentityOptions);
+
+            // Lock the potential write of the dictionary to prevent multiple threads from creating the same application.
+            await _managedIdSemaphore.WaitAsync();
+            try
+            {
+                // Check if the application is already built and not null, if so return it
+                if (_managedIdentityApplicationsByClientId.TryGetValue(key, out application) && application != null)
```

Comment:
not quite a standard double check lock, is it supposed to be?

If it is, check for the presence before trying to acquire the lock.

Reply:
Wasn't familiar with that pattern, thanks!